File: backend/core/src/modules/characters/characters.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { ProjectAccessService } from '../../common/access/project-access.service';
import { ProjectScopedCrudService } from '../../common/access/project-scoped-crud.service';
import { Character } from '../../entities';
import { SearchService } from '../search/search.service';

@Injectable()
export class CharactersService extends ProjectScopedCrudService<Character> {
  constructor(
    @InjectRepository(Character)
    repo: Repository<Character>,
    private readonly searchService: SearchService,
    projectAccessService: ProjectAccessService,
  ) {
    super(repo, projectAccessService);
  }

  override async createForUser(userId: string | undefined, dto: Partial<Character>): Promise<Character> {
    const character = await super.createForUser(userId, dto);
    await this.searchService.indexDocument('characters', character.id, {
      id: character.id,
      projectId: character.projectId,
      name: character.name,
      role: character.role,
      backstory: character.backstory,
    });
    return character;
  }

  override async updateForUser(userId: string | undefined, id: string, dto: Partial<Character>): Promise<Character | null> {
    const character = await super.updateForUser(userId, id, dto);
    if (character) {
      await this.searchService.indexDocument('characters', character.id, {
        id: character.id,
        projectId: character.projectId,
        name: character.name,
        role: character.role,
        backstory: character.backstory,
      });
    }
    return character;
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/service/CharactersService.java
package store.pjcloud.cowrite.core.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.Map;
import java.util.UUID;
import org.springframework.stereotype.Service;
import store.pjcloud.cowrite.core.common.ProjectScopedCrudService;
import store.pjcloud.cowrite.core.entity.Character;
import store.pjcloud.cowrite.core.repository.CharacterRepository;

@Service
public class CharactersService extends ProjectScopedCrudService<Character> {
    private final SearchService searchService;

    public CharactersService(CharacterRepository repository,
                             ProjectAccessService projectAccessService,
                             ObjectMapper objectMapper,
                             SearchService searchService) {
        super(repository, projectAccessService, objectMapper, Character.class);
        this.searchService = searchService;
    }

    @Override
    public Character createForUser(UUID userId, Map<String, Object> body) {
        Character character = super.createForUser(userId, body);
        index(character);
        return character;
    }

    @Override
    public Character updateForUser(UUID userId, UUID id, Map<String, Object> body) {
        Character character = super.updateForUser(userId, id, body);
        if (character != null) index(character);
        return character;
    }

    private void index(Character character) {
        Map<String, Object> doc = new java.util.HashMap<>();
        doc.put("id", character.getId() == null ? null : character.getId().toString());
        doc.put("projectId", character.getProjectId() == null ? null : character.getProjectId().toString());
        doc.put("name", character.getName());
        doc.put("role", character.getRole());
        doc.put("backstory", character.getBackstory());
        searchService.indexDocument("characters", character.getId().toString(), doc);
    }
}
